File: src/lib/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { Session, RawEvent, DbMode, Mode, KernelConfig } from "./types";

// -- Sessions --

export function createSession(projectPath: string) {
  return invoke<Session>("create_session", { projectPath });
}

export function getSession(id: string) {
  return invoke<Session | null>("get_session", { id });
}

export function listSessions() {
  return invoke<Session[]>("list_sessions");
}

export function deleteSession(id: string) {
  return invoke<void>("delete_session", { id });
}

// -- Events --

export function insertEvent(
  sessionId: string,
  agentId: string | null,
  kind: string,
  data: string,
) {
  return invoke<RawEvent>("insert_event", { sessionId, agentId, kind, data });
}

export function eventsSince(sessionId: string, since: string) {
  return invoke<RawEvent[]>("events_since", { sessionId, since });
}

// -- DB Modes --

export function listDbModes() {
  return invoke<DbMode[]>("list_db_modes");
}

export function getDbMode(name: string) {
  return invoke<DbMode | null>("get_db_mode", { name });
}

// -- Config --

export function loadProjectConfig(projectRoot: string) {
  return invoke<KernelConfig>("load_project_config", { projectRoot });
}

// -- Builtin Modes --

export function getBuiltinModes() {
  return invoke<Mode[]>("get_builtin_modes");
}

// -- Prompt Router --

/** The raw conversation context the LLM sees (for debugging compaction). */
export interface ContextMessage {
  role: "user" | "assistant";
  content: ContextBlock[];
}

export type ContextBlock =
  | { type: "text"; text: string }
  | { type: "tool_use"; id: string; name: string; input: Record<string, unknown> }
  | { type: "tool_result"; tool_use_id: string; content: string; is_error: boolean };

export function getConversationContext(sessionId: string) {
  return invoke<ContextMessage[]>("get_conversation_context", { sessionId });
}

// -- Conversation History --

export interface HistoryEntry {
  type: "message" | "interrupted" | "compaction";
  // message fields (when type === "message"):
  role?: "user" | "assistant";
  content?: ContextBlock[];
  // compaction fields (when type === "compaction"):
  before_messages?: number;
  after_messages?: number;
}

export interface ConversationHistory {
  entries: HistoryEntry[];
  last_mode: { mode: string; model: string; confidence: number } | null;
}

export function getConversationHistory(sessionId: string) {
  return invoke<ConversationHistory>("get_conversation_history", { sessionId });
}

export function submitPrompt(
  sessionId: string,
  prompt: string,
  modeOverride: string | null,
) {
  return invoke<void>("submit_prompt", { sessionId, prompt, modeOverride });
}

export function cancelPrompt(sessionId: string) {
  return invoke<void>("cancel_prompt", { sessionId });
}

// -- Session Cost --

export function getSessionCost(sessionId: string): Promise<number> {
  return invoke<number>("get_session_cost", { sessionId });
}

// -- Attached Plan --

export function getAttachedPlan(sessionId: string): Promise<string | null> {
  return invoke<string | null>("get_attached_plan", { sessionId });
}

// -- File Revert --

export type RevertResult =
  | { status: "success" }
  | { status: "conflict"; expected_hash: string; actual_hash: string }
  | { status: "not_found" }
  | { status: "error"; message: string };

export function revertFile(
  sessionId: string,
  toolUseId: string,
  path: string,
  beforeContent: string | null,
  afterContent: string,
  reason: string,
  force?: boolean,
) {
  return invoke<RevertResult>("revert_file", {
    sessionId,
    toolUseId,
    path,
    beforeContent,
    afterContent,
    reason,
    force: force ?? false,
  });
}

File: src-tauri/src/db/commands.rs
use sqlx::SqlitePool;
use tauri::State;
use tracing::{debug, error, info, instrument};

use super::models::{Event, Mode, Session};
use super::queries;

#[tauri::command]
#[instrument(skip(pool))]
pub async fn create_session(
    pool: State<'_, SqlitePool>,
    project_path: String,
) -> Result<Session, String> {
    info!(project_path, "cmd: create_session");
    queries::create_session(&pool, &project_path)
        .await
        .map_err(|e| { error!(error = %e, "create_session failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn get_session(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<Option<Session>, String> {
    debug!(id, "cmd: get_session");
    queries::get_session(&pool, &id)
        .await
        .map_err(|e| { error!(error = %e, "get_session failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn list_sessions(pool: State<'_, SqlitePool>) -> Result<Vec<Session>, String> {
    debug!("cmd: list_sessions");
    queries::list_sessions(&pool)
        .await
        .map_err(|e| { error!(error = %e, "list_sessions failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn delete_session(
    pool: State<'_, SqlitePool>,
    id: String,
) -> Result<(), String> {
    info!(id, "cmd: delete_session");
    queries::delete_session(&pool, &id)
        .await
        .map_err(|e| { error!(error = %e, "delete_session failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn list_db_modes(pool: State<'_, SqlitePool>) -> Result<Vec<Mode>, String> {
    debug!("cmd: list_db_modes");
    queries::list_modes(&pool)
        .await
        .map_err(|e| { error!(error = %e, "list_db_modes failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn get_db_mode(
    pool: State<'_, SqlitePool>,
    name: String,
) -> Result<Option<Mode>, String> {
    debug!(name, "cmd: get_db_mode");
    queries::get_mode(&pool, &name)
        .await
        .map_err(|e| { error!(error = %e, "get_db_mode failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool, data))]
pub async fn insert_event(
    pool: State<'_, SqlitePool>,
    session_id: String,
    agent_id: Option<String>,
    kind: String,
    data: String,
) -> Result<Event, String> {
    debug!(kind, session_id, "cmd: insert_event");
    queries::insert_event(&pool, &session_id, agent_id.as_deref(), &kind, &data)
        .await
        .map_err(|e| { error!(error = %e, "insert_event failed"); e.to_string() })
}

#[tauri::command]
#[instrument(skip(pool))]
pub async fn events_since(
    pool: State<'_, SqlitePool>,
    session_id: String,
    since: String,
) -> Result<Vec<Event>, String> {
    debug!(session_id, since, "cmd: events_since");
    queries::events_since(&pool, &session_id, &since)
        .await
        .map_err(|e| { error!(error = %e, "events_since failed"); e.to_string() })
}
